homeassistant/components/group/entity.py

```diff
@@ -425,27 +440,73 @@ def _async_update_group_state(self, tr_state: State | None = None) -> None:
         elif tr_state.attributes.get(ATTR_ASSUMED_STATE):
             self._assumed_state = True
 
-        num_on_states = len(self._on_states)
+        group_is_on = self.mode(self._on_off.values())
+        # If we do not have an on state for any domains
+        # we use None (which will be STATE_UNKNOWN)
+        if (num_on_states := len(self._on_states)) == 0:
+            self._state = None
+            return
+
+        registry: GroupIntegrationRegistry = self.hass.data[REG_KEY]
+
+        # In case the group contains entities of the same domain with the same ON
+        # or an OFF state (one or more domains), we want to use that specific state.
+        # If we have more then one ON or OFF state we default to STATE_ON or STATE_OFF.
+        active_on_states: set[str] = set()
+        active_off_states: set[str] = set()
+
+        def _detect_specific_on_off_state() -> None:
```

Comment:
I would make this a bound function and call it as `self._detect_specific_on_off_state(registry)` since there are cases were we never call it